From 1d18bb5b0d418e25c1f97d2121b3dd58140f66fe Mon Sep 17 00:00:00 2001
From: Martin Jansa <martin.jansa@lge.com>
Date: Fri, 1 Apr 2022 20:12:37 +0000
Subject: [PATCH] google: link with libgrpc as well

:Release Notes:
Since grpc-1.45 grpc_library.h is included in the build and causes
grpc_init and grpc_symbols to be referenced in GoogleTTSEngine.cpp.o
which results in tts-service linking to fail.

:Detailed Notes:
With new grpc-1.45 introduced in:
https://git.openembedded.org/meta-openembedded/commit/?id=df5b764f99dbb19048ebef9d6b85e4d3625b4272

grpc_library.h is included through tls_credentials_options.h -> tls_certificate_verifier.h since:
https://github.com/krestofur/grpc/commit/2e14f6fa709e3900ea9cf77355b05970edc265af#diff-1154485570db809f9e6a620131d41376bc16cb56315d3003f72b62789e952f86R29

there is:
com.webos.service.tts/1.0.0-20-r6/build $ objdump -x CMakeFiles/tts-service.dir/src/engines/tts/google/GoogleTTSEngine.cpp.o | grep grpc_init
0000000000000000         *UND*  0000000000000000 grpc_init
0000000000000001 R_X86_64_PLT32    grpc_init-0x0000000000000004
000000000000004a R_X86_64_PLT32    grpc_init-0x0000000000000004

com.webos.service.tts/1.0.0-20-r6/build $ objdump -x CMakeFiles/tts-service.dir/src/engines/tts/google/GoogleTTSEngine.cpp.o | grep grpc_shutdown
0000000000000000         *UND*  0000000000000000 grpc_shutdown
0000000000001b1b R_X86_64_PLT32    grpc_shutdown-0x0000000000000004
0000000000000001 R_X86_64_PLT32    grpc_shutdown-0x0000000000000004
0000000000000040 R_X86_64_PLT32    grpc_shutdown-0x0000000000000004
0000000000000034 R_X86_64_PLT32    grpc_shutdown-0x0000000000000004
0000000000000091 R_X86_64_PLT32    grpc_shutdown-0x0000000000000004
0000000000000082 R_X86_64_PLT32    grpc_shutdown-0x0000000000000004

And build fails with:
/usr/include/grpcpp/impl/grpc_library.h:35: error: undefined reference to 'grpc_shutdown'
/usr/include/grpcpp/impl/grpc_library.h:34: error: undefined reference to 'grpc_init'
/usr/include/grpcpp/impl/grpc_library.h:35: error: undefined reference to 'grpc_shutdown'
/usr/include/grpcpp/impl/grpc_library.h:35: error: undefined reference to 'grpc_shutdown'
/usr/include/grpcpp/impl/grpc_library.h:35: error: undefined reference to 'grpc_shutdown'
/usr/include/grpcpp/impl/grpc_library.h:34: error: undefined reference to 'grpc_init'

The symbols are defined only in libgrpc, in libgrpc++ they are undefined as well:
com.webos.service.tts/1.0.0-20-r6/git $ objdump  -T ../recipe-sysroot/usr/lib/libgrpc++.so | grep grpc_shutdown
0000000000000000      DF *UND*  0000000000000000              grpc_shutdown
000000000006e3e0 g    DF .text  0000000000000005  Base        _ZN4grpc11CoreCodegen13grpc_shutdownEv

com.webos.service.tts/1.0.0-20-r6/git $ objdump  -T ../recipe-sysroot/usr/lib/libgrpc.so | grep grpc_shutdown
00000000003cfb50 g    DF .text  0000000000000089  Base        _Z22grpc_shutdown_internalPv
00000000003cfe10 g    DF .text  0000000000000079  Base        grpc_shutdown_blocking
00000000003cf950 g    DF .text  00000000000001fa  Base        _Z29grpc_shutdown_internal_lockedv
00000000003cfbe0 g    DF .text  000000000000022d  Base        grpc_shutdown
0000000000316810 g    DF .text  000000000000000d  Base        _Z32grpc_shutdown_background_closurev

com.webos.service.tts/1.0.0-20-r6/git $ objdump  -T ../recipe-sysroot/usr/lib/libgrpc.so | grep grpc_init
00000000003cf830 g    DF .text  0000000000000120  Base        grpc_init
000000000030dbb0 g    DF .text  000000000000030b  Base        _Z22grpc_init_epoll1_linuxb
00000000003127f0 g    DF .text  0000000000000161  Base        _Z23grpc_init_epollex_linuxb
00000000003158e0 g    DF .text  0000000000000084  Base        _Z20grpc_init_poll_posixb

com.webos.service.tts/1.0.0-20-r6/git $ objdump  -T ../recipe-sysroot/usr/lib/libgrpc++.so | grep grpc_init
0000000000000000      DF *UND*  0000000000000000              grpc_init
000000000006e3d0 g    DF .text  0000000000000005  Base        _ZN4grpc11CoreCodegen9grpc_initEv

Link with both to work around this issue.

:Testing Performed:
Only build tested.

:QA Notes:
No change to image.

:Issues Addressed:
[WRN-9820] Create GPVB with Yocto 3.5 Kirkstone
---
 src/engines/tts/google/CMakeLists.txt | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/engines/tts/google/CMakeLists.txt b/src/engines/tts/google/CMakeLists.txt
index 0510647..f04a47d 100755
--- a/src/engines/tts/google/CMakeLists.txt
+++ b/src/engines/tts/google/CMakeLists.txt
@@ -26,6 +26,7 @@ set(src ${CMAKE_CURRENT_SOURCE_DIR}/GoogleTTSEngine.cpp
     ${GOOGLE_SOURCE}
 )
  set(deps
+     grpc
      grpc++
      protobuf
 )
